File: src/main/java/com/kishlaly/ta/analyze/testing/sl/StopLossVolatileKeltnerBottom.java
package com.kishlaly.ta.analyze.testing.sl;

import com.kishlaly.ta.model.SymbolData;
import com.kishlaly.ta.model.indicators.KeltnerJava;
import com.kishlaly.ta.utils.IndicatorUtils;

public class StopLossVolatileKeltnerBottom extends StopLossStrategy {

    public StopLossVolatileKeltnerBottom(Object config) {
        super(config, true);
    }

    @Override
    public double calculate(SymbolData data, int signalIndex) {
        KeltnerJava keltner = IndicatorUtils.buildKeltnerChannels(data.symbol, data.quotes).get(signalIndex);
        int bottomRatio = (int) getConfig();
        double middle = keltner.getMiddle();
        double bottom = keltner.getLow();
        double diff = middle - bottom;
        double ratio = diff / 100 * bottomRatio;
        return middle - ratio;
    }

    @Override
    public String toString() {
        return "SL volatile Keltner " + getConfig() + "% bottom";
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/testing/sl/StopLossVolatileKeltnerBottom.kt
package com.kishlaly.ta.analyze.testing.sl

import com.kishlaly.ta.model.SymbolData
import com.kishlaly.ta.utils.IndicatorUtils

class StopLossVolatileKeltnerBottom(config: Any?) : StopLossStrategy(config, true) {

    override fun calculate(data: SymbolData, signalIndex: Int): Double {
        val keltner = IndicatorUtils.buildKeltnerChannels(data.symbol, data.quotes)[signalIndex]
        val bottomRatio = config as Int
        val middle = keltner.middle
        val bottom = keltner.low
        val diff = middle - bottom
        val ratio = diff / 100 * bottomRatio
        return middle - ratio
    }

    override fun toString() = "SL volatile Keltner ${config as Int}% bottom"

}